clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/DeployCloudFormationAtomicOperation.java

```diff
@@ -0,0 +1,115 @@
+import com.amazonaws.services.cloudformation.model.*;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.netflix.spinnaker.clouddriver.aws.deploy.description.DeployCloudFormationDescription;
+import com.netflix.spinnaker.clouddriver.aws.security.AmazonClientProvider;
+import com.netflix.spinnaker.clouddriver.data.task.Task;
+import com.netflix.spinnaker.clouddriver.data.task.TaskRepository;
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperation;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class DeployCloudFormationAtomicOperation implements AtomicOperation<Map> {
+
+  private static final String BASE_PHASE = "DEPLOY_CLOUDFORMATION";
+
+
+  @Autowired
+  AmazonClientProvider amazonClientProvider;
+
+  @Autowired
+  private ObjectMapper objectMapper;
+
+  private DeployCloudFormationDescription description;
+
+  public DeployCloudFormationAtomicOperation(DeployCloudFormationDescription deployCloudFormationDescription) {
+    this.description = deployCloudFormationDescription;
+  }
+
+  @Override
+  public Map operate(List priorOutputs) {
+    Task task = TaskRepository.threadLocalTask.get();
+    task.updateStatus(BASE_PHASE, "Configuring cloudformation");
+    AmazonCloudFormation amazonCloudFormation = amazonClientProvider.getAmazonCloudFormation(
+      description.getCredentials(), description.getRegion());
+    String template;
+    try {
+      template = objectMapper.writeValueAsString(description.getTemplateBody());
+    } catch (JsonProcessingException e) {
+      throw new IllegalArgumentException("Could not deserialize template body", e);
+    }
+    List<Parameter> parameters = description.getParameters().entrySet().stream()
+      .map(entry -> new Parameter()
+        .withParameterKey(entry.getKey())
+        .withParameterValue(entry.getValue()))
+      .collect(Collectors.toList());
+    try {
+      String stackId = createStack(amazonCloudFormation, template, parameters);
+      return Collections.singletonMap("stackId", stackId);
+    } catch (AlreadyExistsException e) {
+      String stackId = updateStack(amazonCloudFormation, template, parameters);
+      return Collections.singletonMap("stackId", stackId);
+    }
+  }
+
+  private String createStack(AmazonCloudFormation amazonCloudFormation, String template, List<Parameter> parameters) {
+    Task task = TaskRepository.threadLocalTask.get();
+    task.updateStatus(BASE_PHASE, "Preparing cloudformation");
+    CreateStackRequest createStackRequest = new CreateStackRequest()
+      .withStackName(description.getStackName())
+      .withParameters(parameters)
+      .withTemplateBody(template);
+    task.updateStatus(BASE_PHASE, "Uploading cloudformation");
```

Comment:
I'd recommend changing these to say "[blah blah] CloudFormation Stack" instead.

Reply:
good point, updated 👍